### PR TITLE
Better windows support

### LINKED ISSUE #590: The process cannot access the file because it is being used by another process (windows)
Hi, I consistently get the following error when developing in my PC (windows 10):

```
    train_dataset = train_dataset.map(convert_to_features, batched=True)
  File "C:\Users\saareliad\AppData\Local\Continuum\miniconda3\envs\py38\lib\site-packages\nlp\arrow_dataset.py", line 970, in map
    shutil.move(tmp_file.name, cache_file_name)
  File "C:\Users\saareliad\AppData\Local\Continuum\miniconda3\envs\py38\lib\shutil.py", line 803, in move
    os.unlink(src)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\saareliad\\.cache\\huggingface\\datasets\\squad\\plain_text\\1.0.0\\408a8fa46a1e2805445b793f1022e743428ca739a34809fce872f0c7f17b44ab\\tmpsau1bep1'

```

### PR DESCRIPTION
**Pull request by lhoestq**
There are a few differences in the behavior of python and pyarrow on windows.

For example there are restrictions when accessing/deleting files that are open

Fix #590

### CODE REVIEW

File: src/datasets/builder.py [hunk 1]
```diff
@@ -174,8 +174,8 @@ def __init__(
         self._cache_dir_root = os.path.expanduser(cache_dir or HF_DATASETS_CACHE)
         self._cache_dir = self._build_cache_dir()
         lock_path = os.path.join(self._cache_dir_root, self._cache_dir.replace("/", "_") + ".lock")
-        with FileLock(lock_path):
-            if os.path.exists(self._cache_dir):  # check if data exist
+        if os.path.exists(self._cache_dir):  # check if data exist
+            with FileLock(lock_path, self._cache_dir):
```
Comment (thomwolf) [inline src/datasets/builder.py:178]: Are you sure about this?

`lock_path` is a file inside `self._cache_dir_root`, not inside `self._cache_dir`

**Review by thomwolf — COMMENTED** [review]
Not sure about one modification (maybe it's related to the CI failure)

File: src/datasets/builder.py (hunk 1 above)
Comment (lhoestq) [reply src/datasets/builder.py]: I was testing something. It's back to what it was

**Comment by lhoestq** [comment]
This PR is ready :)
It brings official support for windows.

Some tests `AWSDatasetTest` are failing for windows.
This is because I had to fix a few datasets that were not compatible with windows.
These test will pass once they got merged on master :)

File: .circleci/config.yml [hunk 2]
```diff
@@ -26,6 +30,41 @@ jobs:
             - run: pip install pyarrow==1.0.0
             - run: HF_SCRIPTS_VERSION=master python -m pytest -sv ./tests/
 
+
+    run_dataset_script_tests_pyarrow_0p17_WIN:
```
Comment (thomwolf) [inline .circleci/config.yml:34]: Youpi!

**Review by thomwolf — APPROVED** [review]
Honestly I mostly trust you here :)

**Pull request merged**